Adding boxes to SceneGraph

SeanCurtis-TRI opened the pull request:
This adds the box primitive to SceneGraph for both collision and
visualization.

1. Add box spec to `shape_specification.h` and extend reifier interface to
   include the box.
2. Update the proximity engine, rigid body plant bridge, and visualizer
   reification implementation to account for the boxes.
3. Fix collision parameters using FCL -- both solver type and tolerances.
4. Modify unit tests -- introduce "tangent contact" test to test boxes
   against other geometry types. (Small clean up of previous testing code.)
5. Modify sliding_boxes.cc to verify visualization of boxes.

NOTE: The current FCL has *horrible* precision for collision between box
and quadratic surfaces (sphere and cylinder). The tests in
proximity_engine_test.cc specifically calls those out.


```
Category            added  modified  removed  
----------------------------------------------
code                240    18        4        
comments            110    4         4        
blank               47     0         0        
----------------------------------------------
TOTAL               397    22        8   
```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/8776)
<!-- Reviewable:end -->

amcastro-tri commented:
Review status: 0 of 10 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/8776#-LC13or_7Pvt7WDZH42H:-LC13or_7Pvt7WDZH42I:b-zgx05i) (no related file):*
CI barks due to some lint and the removal of a "playback" flag from `sliding_bricks.cc` that however still gets passed in the `BUILD.bazel` file.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8776)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
One of the delays on this PR has been intermittent efforts to address that. I'd pushed a couple of PRs in fcl which, at the end of the day, didn't make things better.  :(  Some of this will be addressed by not having sphere-cube intersection going to GJK. That's just absurd. But, for now, this gets it in with *some* answers.

Also, it's worth noting that the largest error is in normal direction. When a normal is supposed to be [0, 0, 1], and we get [0.07, 0, 0.9975] it's *close* to the right normal, but the error is still large.

---

Review status: 0 of 10 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/8776#-LC13or_7Pvt7WDZH42H:-LC1shV5Ad0qz6bb3rFU:b-e6j8aa) (no related file):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

CI barks due to some lint and the removal of a "playback" flag from `sliding_bricks.cc` that however still gets passed in the `BUILD.bazel` file.
</blockquote></details>

Thanks for flagging that. I'll clean that up as well.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8776#-:-LC1s6O60QgrEn8S0v1S:b-f7pg1)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
The pose of your cube is wrong and that is why these tests fail. I pulled your PR and tried the new pose that I posted below and the tests pass with very tight tolerances. I have to say that getting that pose right is not that trivial, it took some mental effort.

---

Reviewed 7 of 10 files at r1.
Review status: 7 of 10 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[examples/contact_model/sliding_bricks.cc, line 122 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8776#-LC4PevG3fwjzG8qGH_c:-LC4PevG3fwjzG8qGH_d:b-pdw1pj) ([raw file](https://github.com/robotlocomotion/drake/blob/d3bfa71064a553e5d5131242e4fc37a75bd387d1/examples/contact_model/sliding_bricks.cc#L122)):*
> ```cpp
>   // TODO(SeanCurtis-TRI): This should be wrapped up into a sugar diagram so
>   // this can be done more concisely. Perhaps bundle a SceneGraph in with
>   // all the others and called (something like) SceneGraphWithLcmVisuals.
> ```

Do you mean wrapping it up with the `PoseBundleToDrawMessage` and the `LcmPublisherSystem`? that'd be awesome. Maybe even more than a specialized diagram a method that builds it for us? like `SceneGraph& AddSceneGraphWithLcmVisuals(Builder*, DrakeLcm*)`?

---

*[examples/contact_model/sliding_bricks.cc, line 161 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8776#-LC4QIdt0Gy1txFd6Kqn:-LC4QIdt0Gy1txFd6Kqo:b-xieogo) ([raw file](https://github.com/robotlocomotion/drake/blob/ac1b321c075623a4deaca86a9d6df129bc73cd42/examples/contact_model/sliding_bricks.cc#L161)):*
> ```cpp
>   simulator->StepTo(FLAGS_sim_duration);
> 
>   while (FLAGS_playback) visualizer_publisher->ReplayCachedSimulation();
> ```

BTW It'd be cool having this capability somehow with SG. Too difficult to add?

---

*[geometry/test/proximity_engine_test.cc, line 371 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8776#-LC4a3X---cMvRlg3NAS:-LC4a3X---cMvRlg3NAT:bz36dza) ([raw file](https://github.com/robotlocomotion/drake/blob/d3bfa71064a553e5d5131242e4fc37a75bd387d1/geometry/test/proximity_engine_test.cc#L371)):*
> ```cpp
> // This performs a very specific test. It collides a rotated box with a
> // surface that is tangent to the z = 0 plane. The box is a cube with unit size.
> // It is oriented and positioned so that a single corner, placed on the z axis,
> ```

humm... "a single corner, placed on the z axis," how do you place a corner (a point) on an axis?

I guess you mean so that the diagonal line (of size sqrt(3)) connecting two opposite corners of the cube aligns with the z axis?

---

*[geometry/test/proximity_engine_test.cc, line 399 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8776#-LC4bNXJ4dUQlP2s5yRi:-LC4bNXJ4dUQlP2s5yRj:bff37ap) ([raw file](https://github.com/robotlocomotion/drake/blob/d3bfa71064a553e5d5131242e4fc37a75bd387d1/geometry/test/proximity_engine_test.cc#L399)):*
> ```cpp
> //   - the penetration depth is the specified `depth` used to configure the
> //     position of the box, and
> //   - the contact position is (0, 0, -depth / 2.
> ```

BTW missing closing parenthesis

---

*[geometry/test/proximity_engine_test.cc, line 401 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8776#-LC4beoF1u8ANqvphlif:-LC4beoF1u8ANqvphlig:bn2u4i2) ([raw file](https://github.com/robotlocomotion/drake/blob/d3bfa71064a553e5d5131242e4fc37a75bd387d1/geometry/test/proximity_engine_test.cc#L401)):*
> ```cpp
> //   - the contact position is (0, 0, -depth / 2.
> //
> // Every convex shape type can be used as the tangent shape as follows:
> ```

BTW actually concave shapes as well as long as there is only deepest contact at this corner with the cube.

---

*[geometry/test/proximity_engine_test.cc, line 423 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8776#-LC4cD5s5KAo4nlL6ZeW:-LC4cD5s5KAo4nlL6ZeX:b9qe0ns) ([raw file](https://github.com/robotlocomotion/drake/blob/d3bfa71064a553e5d5131242e4fc37a75bd387d1/geometry/test/proximity_engine_test.cc#L423)):*
> ```cpp
>     // a. Orient the box so that the corner at -x, -y, -z lies in the most -z
>     //    extent.
>     box_pose_ = AngleAxisd(-M_PI_4, Vector3d::UnitY()) *
> ```

Lets use monogram notation? Define a world frame W and a box frame B (additionally if you want frame A for the tangent geometry with X_WA = Identity). Then this would be `X_WB`

---

*[geometry/test/proximity_engine_test.cc, line 423 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8776#-LC4maUh1aUPgYnh51KY:-LC4maUh1aUPgYnh51KZ:byrwd7q) ([raw file](https://github.com/robotlocomotion/drake/blob/d3bfa71064a553e5d5131242e4fc37a75bd387d1/geometry/test/proximity_engine_test.cc#L423)):*
> ```cpp
>     // a. Orient the box so that the corner at -x, -y, -z lies in the most -z
>     //    extent.
>     box_pose_ = AngleAxisd(-M_PI_4, Vector3d::UnitY()) *
> ```

Maybe a VC better, but if I understand correctly, don't you want this to be:
```c++
X_WB.linear() = AngleAxisd(std::atan(M_SQRT2), Vector3d(M_SQRT1_2, -M_SQRT1_2, 0.0)).toRotationMatrix();
```
?

---

*[geometry/test/proximity_engine_test.cc, line 427 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8776#-LC4h6j-BgH6kdaxcSgF:-LC4h6j-BgH6kdaxcSgG:bfaegvd) ([raw file](https://github.com/robotlocomotion/drake/blob/d3bfa71064a553e5d5131242e4fc37a75bd387d1/geometry/test/proximity_engine_test.cc#L427)):*
> ```cpp
> 
>     // b. Translate it so that the transformed corner lies at (0, 0, -d).
>     Vector3d corner{-0.5, -0.5, -0.5};  // The colliding box has unit length.
> ```

Monogram notation? with C the corner point this'd be `p_BoC_B`.

---

*[geometry/test/proximity_engine_test.cc, line 428 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8776#-LC4hDiP3k9JJ8-DJYT9:-LC4hDiP3k9JJ8-DJYTA:bqt6jet) ([raw file](https://github.com/robotlocomotion/drake/blob/d3bfa71064a553e5d5131242e4fc37a75bd387d1/geometry/test/proximity_engine_test.cc#L428)):*
> ```cpp
>     // b. Translate it so that the transformed corner lies at (0, 0, -d).
>     Vector3d corner{-0.5, -0.5, -0.5};  // The colliding box has unit length.
>     Vector3d rotated_corner = box_pose_.linear() * corner;
> ```

Monogram notation? `p_BoC_W = X_WB * p_BoC_B`?

---

*[geometry/test/proximity_engine_test.cc, line 429 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8776#-LC4jaAc9nzanGKXXATI:-LC4jaAc9nzanGKXXATJ:brgtvn0) ([raw file](https://github.com/robotlocomotion/drake/blob/d3bfa71064a553e5d5131242e4fc37a75bd387d1/geometry/test/proximity_engine_test.cc#L429)):*
> ```cpp
>     Vector3d corner{-0.5, -0.5, -0.5};  // The colliding box has unit length.
>     Vector3d rotated_corner = box_pose_.linear() * corner;
>     Vector3d offset = -rotated_corner + Vector3d{0, 0, -kDepth};
> ```

`p_WB = -p_BoC_W + p_WC` (or equivalently `p_WB = p_CBo_W + p_WC`)

with `p_WC =  Vector3d{0, 0, -kDepth}`

---

*[geometry/test/proximity_engine_test.cc, line 429 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8776#-LC4nPI33lnUFHXahQkv:-LC4nPI33lnUFHXahQkw:bhty812) ([raw file](https://github.com/robotlocomotion/drake/blob/d3bfa71064a553e5d5131242e4fc37a75bd387d1/geometry/test/proximity_engine_test.cc#L429)):*
> ```cpp
>     Vector3d corner{-0.5, -0.5, -0.5};  // The colliding box has unit length.
>     Vector3d rotated_corner = box_pose_.linear() * corner;
>     Vector3d offset = -rotated_corner + Vector3d{0, 0, -kDepth};
> ```

I believe you want `offset` (which should be renamed to `p_WBo`) to equal `(0, 0, sqrt(3)/2)`? 

---

*[geometry/test/proximity_engine_test.cc, line 430 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8776#-LC4l4nV3iJAOG7-vTY8:-LC4l4nV3iJAOG7-vTY9:b-a1z96y) ([raw file](https://github.com/robotlocomotion/drake/blob/d3bfa71064a553e5d5131242e4fc37a75bd387d1/geometry/test/proximity_engine_test.cc#L430)):*
> ```cpp
>     Vector3d rotated_corner = box_pose_.linear() * corner;
>     Vector3d offset = -rotated_corner + Vector3d{0, 0, -kDepth};
>     box_pose_.translation() = offset;
> ```

so that this would read `X_WB.translation() = p_WBo` as it should be.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8776#-:-LC4P4_mEO-7_lIt2uMz:b-jq61zy)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
From Slack discussion. The pose I chose makes the problem symmetric and apparently the query works only in that case. @seancurtis-tri's original test is great in that it exposes a problem with this query.

---

Review status: 7 of 10 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[geometry/test/proximity_engine_test.cc, line 371 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8776#-LC4a3X---cMvRlg3NAS:-LC5FjS46HmVCHfoV47k:bpmxwwf) ([raw file](https://github.com/robotlocomotion/drake/blob/d3bfa71064a553e5d5131242e4fc37a75bd387d1/geometry/test/proximity_engine_test.cc#L371)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

humm... "a single corner, placed on the z axis," how do you place a corner (a point) on an axis?

I guess you mean so that the diagonal line (of size sqrt(3)) connecting two opposite corners of the cube aligns with the z axis?
</blockquote></details>

I would say what you told me in Slack. You want the deepest point to be at `p_WC`. Now you choose an orientation `R_WB` of the box so that all other points are outside the other shape and C still is the deepest point (important to make the point that any pose that satisfies this should lead to the same result). You do know the position of the corner in the box frame `p_BoC_B = (0.5, 0.5, 0.5)`. From that information you compute the position of the box's center as `p_WBo_W = p_WC + p_CBo_W = p_WC + R_WB * p_CBo_B =  p_WC - R_WB * p_BoC_B`.
Now you have the pose `X_WB` with `X_WB.linear() = R_WB` and `X_WB.translation() = p_WBo_W`.

---

*[geometry/test/proximity_engine_test.cc, line 423 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8776#-LC4maUh1aUPgYnh51KY:-LC5FLAUAxwNPiA0ZG6L:b51i039) ([raw file](https://github.com/robotlocomotion/drake/blob/d3bfa71064a553e5d5131242e4fc37a75bd387d1/geometry/test/proximity_engine_test.cc#L423)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Maybe a VC better, but if I understand correctly, don't you want this to be:
```c++
X_WB.linear() = AngleAxisd(std::atan(M_SQRT2), Vector3d(M_SQRT1_2, -M_SQRT1_2, 0.0)).toRotationMatrix();
```
?
</blockquote></details>

resolved per Slack. However I think a good idea to also add a test with this pose as a way to document our findings.

---

*[geometry/test/proximity_engine_test.cc, line 429 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8776#-LC4nPI33lnUFHXahQkv:-LC5FYU-5ebPd7iJVpkh:b-l5kq0y) ([raw file](https://github.com/robotlocomotion/drake/blob/d3bfa71064a553e5d5131242e4fc37a75bd387d1/geometry/test/proximity_engine_test.cc#L429)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

I believe you want `offset` (which should be renamed to `p_WBo`) to equal `(0, 0, sqrt(3)/2)`? 
</blockquote></details>

Resolved in Slack discussion.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8776#-:-LC5EuBqDWG7LAcWz-1I:bk0vnuf)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> . Just small notation issues and a confusion I had when reading the docs.

---

Reviewed 2 of 10 files at r1.
Review status: 9 of 10 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8776#-:-LCA8_3c7u-4NFPuERPY:bq78etk)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Review status: 5 of 10 files reviewed at latest revision, 10 unresolved discussions.

---

*[examples/contact_model/sliding_bricks.cc, line 122 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8776#-LC4PevG3fwjzG8qGH_c:-LC9rUxS0PyHtB9xUmva:bom2o6a) ([raw file](https://github.com/robotlocomotion/drake/blob/d3bfa71064a553e5d5131242e4fc37a75bd387d1/examples/contact_model/sliding_bricks.cc#L122)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Do you mean wrapping it up with the `PoseBundleToDrawMessage` and the `LcmPublisherSystem`? that'd be awesome. Maybe even more than a specialized diagram a method that builds it for us? like `SceneGraph& AddSceneGraphWithLcmVisuals(Builder*, DrakeLcm*)`?
</blockquote></details>

This isn't the venue for this discussion; so I removed the brainstorming and reduced the TODO to the main point.

---

*[examples/contact_model/sliding_bricks.cc, line 161 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8776#-LC4QIdt0Gy1txFd6Kqn:-LC9ra429ZCEo6EIr3Wn:bt9h00g) ([raw file](https://github.com/robotlocomotion/drake/blob/ac1b321c075623a4deaca86a9d6df129bc73cd42/examples/contact_model/sliding_bricks.cc#L161)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW It'd be cool having this capability somehow with SG. Too difficult to add?
</blockquote></details>

OK That could eventually be rolled into whatever the visualization sugar becomes.

---

*[geometry/test/proximity_engine_test.cc, line 371 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8776#-LC4a3X---cMvRlg3NAS:-LC9wWjo5RoWzQxBSU3A:b-vr327c) ([raw file](https://github.com/robotlocomotion/drake/blob/d3bfa71064a553e5d5131242e4fc37a75bd387d1/geometry/test/proximity_engine_test.cc#L371)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

I would say what you told me in Slack. You want the deepest point to be at `p_WC`. Now you choose an orientation `R_WB` of the box so that all other points are outside the other shape and C still is the deepest point (important to make the point that any pose that satisfies this should lead to the same result). You do know the position of the corner in the box frame `p_BoC_B = (0.5, 0.5, 0.5)`. From that information you compute the position of the box's center as `p_WBo_W = p_WC + p_CBo_W = p_WC + R_WB * p_CBo_B =  p_WC - R_WB * p_BoC_B`.
Now you have the pose `X_WB` with `X_WB.linear() = R_WB` and `X_WB.translation() = p_WBo_W`.
</blockquote></details>

I've modified the text to reflect your comment -- it's not as elaborate as your comment, but I think it captures what's needed in this comment. PTAL

---

*[geometry/test/proximity_engine_test.cc, line 399 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8776#-LC4bNXJ4dUQlP2s5yRi:-LC9uXKr3F3o9-V4HuNb:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/d3bfa71064a553e5d5131242e4fc37a75bd387d1/geometry/test/proximity_engine_test.cc#L399)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW missing closing parenthesis
</blockquote></details>

Done

---

*[geometry/test/proximity_engine_test.cc, line 401 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8776#-LC4beoF1u8ANqvphlif:-LC9ubk46NBrIrxsb_DR:bzgd8h3) ([raw file](https://github.com/robotlocomotion/drake/blob/d3bfa71064a553e5d5131242e4fc37a75bd387d1/geometry/test/proximity_engine_test.cc#L401)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW actually concave shapes as well as long as there is only deepest contact at this corner with the cube.
</blockquote></details>

OK Going down that road will require a *bunch* of caveats and clarifications. So, I'm going to leave this as convex and keep it simple.

---

*[geometry/test/proximity_engine_test.cc, line 423 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8776#-LC4cD5s5KAo4nlL6ZeW:-LC9uqMxEi-gdmnPIReP:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/d3bfa71064a553e5d5131242e4fc37a75bd387d1/geometry/test/proximity_engine_test.cc#L423)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Lets use monogram notation? Define a world frame W and a box frame B (additionally if you want frame A for the tangent geometry with X_WA = Identity). Then this would be `X_WB`
</blockquote></details>

Done

---

*[geometry/test/proximity_engine_test.cc, line 423 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8776#-LC4maUh1aUPgYnh51KY:-LCA5hg_D3HjK-4X4wON:b-oexso9) ([raw file](https://github.com/robotlocomotion/drake/blob/d3bfa71064a553e5d5131242e4fc37a75bd387d1/geometry/test/proximity_engine_test.cc#L423)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

resolved per Slack. However I think a good idea to also add a test with this pose as a way to document our findings.
</blockquote></details>

Done I agree, both tests illustrate something of value. I've extended the test to do both.

---

*[geometry/test/proximity_engine_test.cc, line 427 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8776#-LC4h6j-BgH6kdaxcSgF:-LC9uvPAECzv9J70mVZK:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/d3bfa71064a553e5d5131242e4fc37a75bd387d1/geometry/test/proximity_engine_test.cc#L427)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Monogram notation? with C the corner point this'd be `p_BoC_B`.
</blockquote></details>

Done

---

*[geometry/test/proximity_engine_test.cc, line 428 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8776#-LC4hDiP3k9JJ8-DJYT9:-LC9wOp463TMyz4w282Q:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/d3bfa71064a553e5d5131242e4fc37a75bd387d1/geometry/test/proximity_engine_test.cc#L428)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Monogram notation? `p_BoC_W = X_WB * p_BoC_B`?
</blockquote></details>

Done

---

*[geometry/test/proximity_engine_test.cc, line 429 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8776#-LC4jaAc9nzanGKXXATI:-LC9wPSe-_4ylM5T4Iyi:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/d3bfa71064a553e5d5131242e4fc37a75bd387d1/geometry/test/proximity_engine_test.cc#L429)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

`p_WB = -p_BoC_W + p_WC` (or equivalently `p_WB = p_CBo_W + p_WC`)

with `p_WC =  Vector3d{0, 0, -kDepth}`
</blockquote></details>

Done

---

*[geometry/test/proximity_engine_test.cc, line 430 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8776#-LC4l4nV3iJAOG7-vTY8:-LC9wQRy6bRG09UQpAKK:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/d3bfa71064a553e5d5131242e4fc37a75bd387d1/geometry/test/proximity_engine_test.cc#L430)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

so that this would read `X_WB.translation() = p_WBo` as it should be.
</blockquote></details>

Done

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8776)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Comments addressed; PTAL

---

Review status: 5 of 10 files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8776#-:-LCAIDB3EYxj67BlDZ-Y:b-u5gt6o)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
+@ggould-tri for platform review; it can wait til Friday.

---

Review status: 5 of 10 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8776#-:-LCAjl5l4QdkK-n9PHNk:b6ewt5j)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Reviewed 5 of 5 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[geometry/test/proximity_engine_test.cc, line 423 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8776#-LC4maUh1aUPgYnh51KY:-LCAYYfxE9xxEuWTaJtj:b-kaqamb) ([raw file](https://github.com/robotlocomotion/drake/blob/d3bfa71064a553e5d5131242e4fc37a75bd387d1/geometry/test/proximity_engine_test.cc#L423)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

Done I agree, both tests illustrate something of value. I've extended the test to do both.
</blockquote></details>

tanks!

---

*[geometry/test/proximity_engine_test.cc, line 374 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8776#-LCAXTI0AKewRP4rjjYs:-LCAXTI0AKewRP4rjjYt:b-bhsemk) ([raw file](https://github.com/robotlocomotion/drake/blob/ccb75afe1f072a498bc2930b67bc6ff7ccad429c/geometry/test/proximity_engine_test.cc#L374)):*
> ```cpp
> // penetration `p_WC` = (0, 0, -depth). The box is positioned and oriented
> // such that the corner `p_BoC_B` = (-0.5, -0.5, -0.5) is coincident with
> // `p_WC`.  And all other transformed corners lie above the z = 0 plane.
> ```

BTW to be consistent (since a) p_BoC is not the corner but its position and b) p_BoC_B cannot be the same as p_WC since they are in different frames) please add:

"such that the corner point C located at `p_BoC_B = (-0.5, -0.5, -0.5)` in the body frame, is located at `p_WC` in the world frame.  And all other ...."

---

*[geometry/test/proximity_engine_test.cc, line 429 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8776#-LCAYTK35n15q0xSR7AL:-LCAYTK35n15q0xSR7AM:b-5124v0) ([raw file](https://github.com/robotlocomotion/drake/blob/ccb75afe1f072a498bc2930b67bc6ff7ccad429c/geometry/test/proximity_engine_test.cc#L429)):*
> ```cpp
> 
>     // Confirm that the poses of the box satisfy the conditions described above.
>     for (auto func : {X_WB_1, X_WB_2}) {
> ```

BTW, cool!

---

*[geometry/test/proximity_engine_test.cc, line 635 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8776#-LCAbN_9Bs5KlqbNxr4M:-LCAbN_9Bs5KlqbNxr4N:b2eq6b5) ([raw file](https://github.com/robotlocomotion/drake/blob/ccb75afe1f072a498bc2930b67bc6ff7ccad429c/geometry/test/proximity_engine_test.cc#L635)):*
> ```cpp
>   // enforces a ratio of at least 100. Using these in a GTEST precludes the
>   // possibility of being constexpr initialized; GTEST takes references.
>   static const double kDepth;
> ```

BTW `kDepth_`?, same below

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8776)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
One more iteration on the docs

---

Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[geometry/test/proximity_engine_test.cc, line 374 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8776#-LCAXTI0AKewRP4rjjYs:-LCAwy27BqVW2LEYXwVY:b-19gt42) ([raw file](https://github.com/robotlocomotion/drake/blob/ccb75afe1f072a498bc2930b67bc6ff7ccad429c/geometry/test/proximity_engine_test.cc#L374)):*
I can almost see what you're saying and am largely inclined to defer to it.  There's an abstraction in your correction that I haven't fully internalized. However, I feel:

> p_BoC_B in the body frame and p_WC in the world frame are redundant.

In the first case, `_B` is short hand for "in the body frame", and the `p_WC` is implicitly `p_WC_W` with its corresponding interpretation. 

So, I've taken liberties of modifying *my* language to be true to the spirit of what you've written but with which I'm more comfortable. Let me know if I still missed the target.

---

*[geometry/test/proximity_engine_test.cc, line 635 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8776#-LCAbN_9Bs5KlqbNxr4M:-LCAwXCgCy05PYtmiWzC:b-666dxn) ([raw file](https://github.com/robotlocomotion/drake/blob/ccb75afe1f072a498bc2930b67bc6ff7ccad429c/geometry/test/proximity_engine_test.cc#L635)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW `kDepth_`?, same below
</blockquote></details>

OK I'm opting for [this interpretation](http://drake.mit.edu/styleguide/cppguide.html#Constant_Names). In this case, the constness beats out the belonging-to-class-ness.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8776#-:-LCAzr3q7UobBtnbDOwA:bbp2r3w)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 9 of 10 files reviewed at latest revision, 2 unresolved discussions.

---

*[geometry/test/proximity_engine_test.cc, line 374 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8776#-LCAXTI0AKewRP4rjjYs:-LCB2Pl702fdr95Yi_vn:bszelpd) ([raw file](https://github.com/robotlocomotion/drake/blob/ccb75afe1f072a498bc2930b67bc6ff7ccad429c/geometry/test/proximity_engine_test.cc#L374)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

I can almost see what you're saying and am largely inclined to defer to it.  There's an abstraction in your correction that I haven't fully internalized. However, I feel:

> p_BoC_B in the body frame and p_WC in the world frame are redundant.

In the first case, `_B` is short hand for "in the body frame", and the `p_WC` is implicitly `p_WC_W` with its corresponding interpretation. 

So, I've taken liberties of modifying *my* language to be true to the spirit of what you've written but with which I'm more comfortable. Let me know if I still missed the target.
</blockquote></details>

Great, thank you.

---

*[geometry/test/proximity_engine_test.cc, line 635 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8776#-LCAbN_9Bs5KlqbNxr4M:-LCB2sgA8OUmFzriNhu_:baz3azd) ([raw file](https://github.com/robotlocomotion/drake/blob/ccb75afe1f072a498bc2930b67bc6ff7ccad429c/geometry/test/proximity_engine_test.cc#L635)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

OK I'm opting for [this interpretation](http://drake.mit.edu/styleguide/cppguide.html#Constant_Names). In this case, the constness beats out the belonging-to-class-ness.
</blockquote></details>

no, you are absolutely right. 

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8776)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8776)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Reviewed 5 of 10 files at r1, 2 of 5 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8776)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>; notes below.  Broadly speaking, it would be good to file an issue with FCL or to put a note about the FCL problem into a Drake issue so that you can reference an issue number in all of the various TODO comments you've put in around the collision quality issues.

---

Reviewed 2 of 5 files at r2, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[examples/contact_model/sliding_bricks.cc, line 121 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8776#-LCE_YSd4WmxduDQEV3t:-LCE_YSd4WmxduDQEV3u:b-894sir) ([raw file](https://github.com/robotlocomotion/drake/blob/7fd6b7ce1b2b48ad94080f91a1ba316db590eeb9/examples/contact_model/sliding_bricks.cc#L121)):*
> ```cpp
> 
>   // TODO(SeanCurtis-TRI): This should be wrapped up into sugar so that
>   // visualization can be added without all this boilerplate.
> ```

BTW -- consider opening an issue for this and referencing that issue in the todo.

---

*[geometry/proximity_engine.cc, line 167 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8776#-LCER1_Z5wtwUQRMnzhh:-LCER1_Z5wtwUQRMnzhi:bb7829e) ([raw file](https://github.com/robotlocomotion/drake/blob/7fd6b7ce1b2b48ad94080f91a1ba316db590eeb9/geometry/proximity_engine.cc#L167)):*
> ```cpp
>       double depth = contact.penetration_depth;
> ```

BTW excess vertical whitespace.

---

*[geometry/proximity_engine.cc, line 433 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8776#-LCERlG4A8D2V2gT4RbE:-LCERlG4A8D2V2gT4RbF:b3scq4u) ([raw file](https://github.com/robotlocomotion/drake/blob/7fd6b7ce1b2b48ad94080f91a1ba316db590eeb9/geometry/proximity_engine.cc#L433)):*
> ```cpp
>     collision_data.request.num_max_contacts = 1;
>     collision_data.request.enable_contact = true;
>     // TODO(SeanCurtis-TRI): FCL has some *significant* issues with its
> ```

It is not clear to me what the concrete TODO is here.  Do we mean to fix FCL?  Is the TODO to retry this when there are future FCL updates in case they fix this?

---

*[geometry/test/proximity_engine_test.cc, line 508 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8776#-LCEW9D_5pUnJsF0JF6K:-LCEW9D_5pUnJsF0JF6L:b-8y6w6r) ([raw file](https://github.com/robotlocomotion/drake/blob/7fd6b7ce1b2b48ad94080f91a1ba316db590eeb9/geometry/test/proximity_engine_test.cc#L508)):*
> ```cpp
> 
>     ASSERT_EQ(results.size(), 1u)
>                   << "Against tangent " << shape_name(shape_type);
> ```

FYI -- I can't quite discern why this indentation would be correct, and am surprised the linter accepts it.  Please double-check that the linter is operating correctly.

---

*[geometry/test/proximity_engine_test.cc, line 562 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8776#-LCEY2O73tRP9Yfj5h-a:-LCEY2O86NIvKsPbyRl7:b-stfvsb) ([raw file](https://github.com/robotlocomotion/drake/blob/7fd6b7ce1b2b48ad94080f91a1ba316db590eeb9/geometry/test/proximity_engine_test.cc#L562)):*
> ```cpp
> 
>   // Produces the X_WB that produces low-quality answers.
>   static Isometry3d X_WB_1(const Vector3d& p_BoC_B, const Vector3d& p_WC) {
> ```

This is a seriously uninformative function name.  Consider naming it with reference to the issue it is testing (eg, the upstream bug tracker number, or "X_WB_that_breaks_fcl" or something.)

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8776#-:-LCE_lfg0u8ef17QrY_r:b-dm5vhd)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Excellent point. I've got a WIP branch for FCL. For now, I'll make it a Drake issue which will reference the FCL issue/pr when that is ready.

---

Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[examples/contact_model/sliding_bricks.cc, line 121 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8776#-LCE_YSd4WmxduDQEV3t:-LCEhHMa0EMIvbfDQM6L:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/7fd6b7ce1b2b48ad94080f91a1ba316db590eeb9/examples/contact_model/sliding_bricks.cc#L121)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

BTW -- consider opening an issue for this and referencing that issue in the todo.
</blockquote></details>

Done

---

*[geometry/proximity_engine.cc, line 167 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8776#-LCER1_Z5wtwUQRMnzhh:-LCEhLc3DaxIsX0jVeN8:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/7fd6b7ce1b2b48ad94080f91a1ba316db590eeb9/geometry/proximity_engine.cc#L167)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

BTW excess vertical whitespace.
</blockquote></details>

Done

---

*[geometry/proximity_engine.cc, line 433 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8776#-LCERlG4A8D2V2gT4RbE:-LCEhvNGBC5oG7V-Anze:b-ociaj8) ([raw file](https://github.com/robotlocomotion/drake/blob/7fd6b7ce1b2b48ad94080f91a1ba316db590eeb9/geometry/proximity_engine.cc#L433)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

It is not clear to me what the concrete TODO is here.  Do we mean to fix FCL?  Is the TODO to retry this when there are future FCL updates in case they fix this?
</blockquote></details>

Moved away from a TODO and more a note of explanation.

---

*[geometry/test/proximity_engine_test.cc, line 508 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8776#-LCEW9D_5pUnJsF0JF6K:-LCEiCdMB8m3ghDn1zzu:bk3pgf6) ([raw file](https://github.com/robotlocomotion/drake/blob/7fd6b7ce1b2b48ad94080f91a1ba316db590eeb9/geometry/test/proximity_engine_test.cc#L508)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

FYI -- I can't quite discern why this indentation would be correct, and am surprised the linter accepts it.  Please double-check that the linter is operating correctly.
</blockquote></details>

Done Some clang format made it happier.

---

*[geometry/test/proximity_engine_test.cc, line 562 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8776#-LCEY2O73tRP9Yfj5h-a:-LCEiF3S78la4V-0O7E4:b-2nywia) ([raw file](https://github.com/robotlocomotion/drake/blob/7fd6b7ce1b2b48ad94080f91a1ba316db590eeb9/geometry/test/proximity_engine_test.cc#L562)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

This is a seriously uninformative function name.  Consider naming it with reference to the issue it is testing (eg, the upstream bug tracker number, or "X_WB_that_breaks_fcl" or something.)
</blockquote></details>

Funny you should say that. I had originally called them `X_WB_good` and `X_WB_bad`. But then I thought, when FCL is fixed, they'll both be good. And, as it turns out, even the "good" one isn't that good in some circumstances (test against the prone cylinder). So, I'm left feeling somewhat ambivalent.

My solution was to also add reference to the drake issue in the note discussing the different poses.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8776#-:-LCEfG2VE912VlhSZj21:bkfx0ca)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Reviewed 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8776)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[examples/contact_model/sliding_bricks.cc, line 121 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8776#-LCE_YSd4WmxduDQEV3t:-LCExfdC2TBOhKTHelZ1:by8pr4n) ([raw file](https://github.com/robotlocomotion/drake/blob/7fd6b7ce1b2b48ad94080f91a1ba316db590eeb9/examples/contact_model/sliding_bricks.cc#L121)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

Done
</blockquote></details>

I don't see a change?

---

*[geometry/test/proximity_engine_test.cc, line 562 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8776#-LCEY2O73tRP9Yfj5h-a:-LCExbSfDvt2kHwZtXC7:bb6mt9b) ([raw file](https://github.com/robotlocomotion/drake/blob/7fd6b7ce1b2b48ad94080f91a1ba316db590eeb9/geometry/test/proximity_engine_test.cc#L562)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

Funny you should say that. I had originally called them `X_WB_good` and `X_WB_bad`. But then I thought, when FCL is fixed, they'll both be good. And, as it turns out, even the "good" one isn't that good in some circumstances (test against the prone cylinder). So, I'm left feeling somewhat ambivalent.

My solution was to also add reference to the drake issue in the note discussing the different poses.
</blockquote></details>

:+1:

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8776)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
This time for real.

---

Review status: 9 of 10 files reviewed at latest revision, 1 unresolved discussion.

---

*[examples/contact_model/sliding_bricks.cc, line 121 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8776#-LCE_YSd4WmxduDQEV3t:-LCF2EGACrfzYcASesxX:b-youlf) ([raw file](https://github.com/robotlocomotion/drake/blob/7fd6b7ce1b2b48ad94080f91a1ba316db590eeb9/examples/contact_model/sliding_bricks.cc#L121)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

I don't see a change?
</blockquote></details>

Done (really this time). Oops...my brain rolled this into the precision issue. 

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8776#-:-LCF2xXe0752UoXh5hya:bubyyoz)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8776)*
<!-- Sent from Reviewable.io -->
